rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,117 @@
+---
+authors: Alex McGrath (alex.mcgrath@goteleport.com
+state: draft
+---
+
+# RFD 57 - Automatic discovery and enrollment of AWS servers
+
+## What
+
+Proposes a way by which an auth service might automatically discover and
+register AWS EC2 instances.
+
+## Why
+
+Currently when adding a new AWS server, its required that Teleport be installed
+after the server has been provisioned which may be a slow process for
+organizations with large numbers of servers as it needs to be installed and then
+added to the teleport cluster
+
+With the changes described in this document, Teleport will be able to resolve
+the issues with adding AWS servers to Teleport clusters automatically.
+
+
+## Discovery
+
+A Teleport auth agent will need to be configured with an assume role policy for
+each of the accounts that the agent will discover services on as described by
+the EC2 join method docs[1]
+
+Discovery can use a matcher similar to the `db_service/aws` matcher:
+
+```yaml
+auth_service:
+  enabled: "yes"
+  aws:
+  - types: ["ec2"]
+    regions: ["us-west-1"]
+    tags:
+      "teleport": "yes"
```

Comment:
In my head this feature will automatically create an EC2 join token, if you are going to filter by tags we may want to extend the EC2 join method to be able to only allow instances with certain tags to join as well, rather than any instance in the account/region (should be pretty doable).